mkosi

```diff
@@ -3147,17 +3151,63 @@ def run_shell(args):
 
         sys.exit(255)
 
+def run_qemu(args):
+
+    if platform.machine() == "x86_64":
+        arch_binary = "qemu-system-x86_64"
+    else:
+        arch_binary = None
+
+    if arch_binary is not None:
+        found = shutil.which(arch_binary)
+    else:
+        found = None
+
+    if found is None:
+        found = shutil.which("qemu")
+
+    if found is None:
+        found = shutil.which("qemu-kvm")
+        if found is None:
+            die("Couldn't find QEMU/KVM binary")
+
+        cmdline = [ found ]
+    else:
+        cmdline = [ found, "-machine", "accel=kvm" ]
+
+    found = None
+    for i in ( '/usr/share/edk2/ovmf/OVMF_CODE.fd', ):
+        if os.path.exists(i):
+            found = i
+
+    if found is None:
+        die("Couldn't find OVMF UEFI firmware blob.")
+
+    cmdline += [ "-bios", found,
+                 "-smp", "2",
+                 "-m", "1024",
+                 "-drive", "format=raw,file=" + args.output,
+                 *args.cmdline ]
+
+    try:
+        subprocess.run(cmdline, check=True)
+    except subprocess.CalledProcessError as e:
+        if e.returncode >= 0:
```

Comment:
`> 0`?

Reply:
we want to propagate the return code including when it is 0... not following here?

Reply:
I don't understand how this is supposed to work. I'd expect that if subprocess.run() is successful, exit(0) is called, and exit(n), where 0 < n <= 255 is called. Calling exit(0) when CalledProcessError was thrown seems wrong. But the docs for CalledProcessError specify that it is only called when the command returns a non-zero exit status, so this check seems redundant anyway.

I'd expect something like this:
```python
sys.exit(subprocess.run(cmdline).returncode)
```
but maybe I'm missing the reason to mangle the return code?

Reply:
if the invoked process fails and aborts due to a signal, then `CalledProcessError.returncode` will be negative. If it exited with a non-zero exit code, it will be positive. In either case we want mkosi to fail with a non-zero exit code. The code you are looking at hence propagates the returncode as is if it is positive, and if it is negative returns 255 (simply because there's no better value to return really, as propagating signals as signals probably doesn't make too much sense...)

Reply:
thinking about this, I figure it would be nicer for the case where we are just supposed to execute qemu or nspawn, to use plain execve() instead of forking something off. After all, we are little more than a preparator for the qemu/nspawn command line we'll eventually invoke.